### PR TITLE
 Support for OpenAPI3

### LINKED ISSUE #84: Support for OpenAPI3
This spans over multiple repos, but can be done in parts.

## Schema -> OpenAPI3
* https://github.com/metosin/schema-tools/issues/44
* might be some work to get this clean without copying a lot

## Spec -> OpenAPI3
* https://github.com/metosin/spec-tools/issues/105
* should be easy, a new ns just like `spec-tools.swagger` overriding some JSON Schema parts (if any!)

## Malli -> OpenAPI3
* https://github.com/metosin/reitit/blob/master/modules/reitit-malli/src/reitit/coercion/malli.cljc

## Reitit Coercion
* currently, there is only one body-coercion for request and one for response. We should add support for content-type based coercion:

### Coercion for all content-types

```clj
(require '[reitit.ring :as ring])
(require '[reitit.ring.coercion])
(require '[reitit.coercion.spec])
(require '[muuntaja.middleware])

(ring/ring-handler
  (ring/router
    ["/plus"
     {:get {:parameters {:body {:x int?, :y int?}}
            :responses {200 {:body {:total pos-int?}}}
            :handler (fn [{{{:keys [x y]} :body} :parameters}]
                       {:status 200, :body (+ x y)})}}]
    {:data {:middleware [muuntaja.middleware/wrap-format
                         reitit.ring.coercion/coerce-request-middleware
                         reitit.ring.coercion/coerce-response-middleware]
            :coercion reitit.coercion.spec/coercion}}))
```

### Coercion per content-type

* new type `:content` (from OpenAPI3) which allows content-type->model mappings.
* should be disallowed with Swagger

```clj
(ring/ring-handler
  (ring/router
    ["/plus"
     {:get {:parameters {:content {"application/json" {:x neg-int?, :y neg-int?}
                                   "application/edn" {:x pos-int?, :y pos-int?}}}
            :responses {200 {:content {"application/json" {:json int?}
                                       "application/edn" {:edn int?}}}}
            :handler (fn [{{{:keys [x y]} :body} :parameters :as request}]
                       (let [content-type (-> request :muuntaja/response :format)
                             body (case content-type
                                    "application/json" {:json (+ x y)}
                                    "application/edn" {:edn (+ x y)})]
                         {:status 200, :body body}))}}]
    {:data {:middleware [muuntaja.middleware/wrap-format
                         reitit.ring.coercion/coerce-request-middleware
                         reitit.ring.coercion/coerce-response-middleware]
            :coercion reitit.coercion.spec/coercion}}))
```

... having both `:body` and `:content` could be allowed, the `:body` could be used as a default if the content-type doesn't match.

### PR DESCRIPTION
**Pull request by souenzzo**
- [x] Schema -> OpenAPI3
- [x] Malli -> OpenAPI3 (using json-schema)
- [x] Spec -> OpenAPI3
- [x] Coercion for all content-types
- [x] Coercion per content-type

Fixes #84

### CODE REVIEW

**Comment by souenzzo** [comment]
Malli is the only one that supports [multiple content types](https://github.com/metosin/reitit/pull/563/files#diff-5796817e1ac259fbdcf73f682599f4bdd3f0422f36e1e93e145c6af202c32344R137)

The API:
- Add `:content-types ["application/json" "application/foo+json"]` to your path/route/router.
- A key `:content-type` will be present in the `option` map, in your transformer.

I'm not sure if this API is enough.

I would like to have a feedback on it.

If this API seems to be enough, I can extend it to schema/spec too.

File: modules/reitit-openapi/src/reitit/openapi.cljc [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+                                   (meta-merge
+                                    #_base-openapi-spec
+                                    (apply meta-merge (keep (comp :openapi :data) middleware))
+                                    (apply meta-merge (keep (comp :openapi :data) interceptors))
+                                    (if coercion
+                                      (coercion/get-apidocs coercion :openapi data))
+                                    (select-keys data [:tags :summary :description])
+                                    (strip-top-level-keys openapi))]))
+           transform-path (fn [[p _ c]]
+                            (if-let [endpoint (some->> c (keep transform-endpoint) (seq) (into {}))]
```
Comment (piotr-yuxuan) [inline modules/reitit-openapi/src/reitit/openapi.cljc:99]:
```suggestion
                            (if-let [endpoint (some->> c (keep transform-endpoint) seq (into {}))]
```

File: modules/reitit-schema/src/reitit/coercion/schema.cljc [hunk 2]
```diff
@@ -67,6 +68,33 @@
...
+                        (empty responses)
+                        (for [[k response] responses]
+                          [k (merge
+                               (select-keys response [:description])
+                               (when (:body response)
+                                 (openapi/openapi-spec
+                                   {::openapi/content {"application/json" (coercion/-compile-model this (:body response) nil)}}))
+                               (when (:content response)
+                                 (openapi/openapi-spec
+                                   {::openapi/content (:content response)})))]))}))
```
Comment (piotr-yuxuan) [inline modules/reitit-schema/src/reitit/coercion/schema.cljc:96]:
```suggestion
                   (when responses
                     (->> responses
                          (reduce-kv (fn [acc k response]
                                       (assoc acc k (merge
                                                      (select-keys response [:description])
                                                      (when (:body response)
                                                        (openapi/openapi-spec
                                                          {::openapi/content {"application/json" (coercion/-compile-model this (:body response) nil)}}))
                                                      (when (:content response)
                                                        (openapi/openapi-spec
                                                          {::openapi/content (:content response)})))))
                                     (empty responses))
                          (assoc {} :responses)))
```

**Review by piotr-yuxuan — APPROVED** [review]
I praise this very good work, and am thankul that you did it! I wanted it so much but shamefully didn't work on it. I haven't tried this to run this diff but it looks quite great. I command your code quality and the extensive work you did 🚀 I quite like that Open API 3.0 and Swagger may coexist alongside.

I'm just a rando an GitHub and have no power on this repo, but I left a couple of minor comments. Take them very freely. Please don't take the style comments very seriously: I didn't pedantically left them, they are more to help me understand and reflect on how to structure code 🙂

**Comment by StankovicMarko** [comment]
any plans on this getting merged?

**Comment by stathissideris** [comment]
Is there any work left on this? I'd be willing to help out.

**Comment by ikitommi** [comment]
Thanks for all the work on this! As this is a big change, planning to make a maintenance release before digging into this.

**Comment by ikitommi** [comment]
Merged with master as thought it was ok but there is a change in malli json schema rendering causing the tests fail now. we can fix those when review, sorry.

**Comment by coyotesqrl** [comment]
This is exciting. I've already had to hack in swagger->openapi on one project and never want to do that again, so being able to autogenerate openapi as easily as swagger would be great.

One thing I noticed on a _very_ brief scan was that there didn't seem to be an analog to `reitit.swagger-ui/create-swagger-ui-handler`. Is that something likely to be added sometime shortly after this is released (hopefully that's soon), or am I seriously underestimating the effort required for that?

**Comment by opqdonut** [comment]
I'm looking at what's needed to be able to merge this. It's a good base. The most notable missing piece is the validation/coercion for the per-content-type models.

**Pull request merged**